### PR TITLE
Fixed modeline width between emacs and emacsclient

### PR DESCRIPTION
**Pull request by Dr-Dd**
Added a condition in which, if an emacs daemon is running in the background, the correct window size is chosen via `window-width`. If a normal emacs istance is running, use instead `window-font-width` as usual. 

Only edge case is if someone opens a standalone emacs istance while running an emacs daemon, but i didn't think it was worth handling it since if someone uses standalone emacs then they are not running the daemon.

### CODE REVIEW

File: doom-modeline-core.el [hunk 1]
```diff
@@ -789,7 +789,8 @@ then this function does nothing."
   "Cache the font width."
   (let ((attributes (face-all-attributes 'mode-line)))
     (or (cdr (assoc attributes doom-modeline--font-width-cache))
-        (let ((width (window-font-width nil 'mode-line)))
+        (let ((width (if (server-running-p) (window-width)
```
Comment (seagle0128) [inline doom-modeline-core.el:792]: Should check if `server-running-p` is loaded and valid.

Comment (Dr-Dd) [reply doom-modeline-core.el:792]: > Should check if server-running-p is loaded and valid.

You mean as in using fboundp?
```suggestion
        (let ((width (if (fboundp 'server-running-p) (window-width)
```
>  And why using (window-width) here for font width?

As i've written in this issue #354, by pure experimentation i found out that the former works with emacs daemon, while the latter (`window-font-width`) with the standalone instance. Don't want to cargo cult here but i really couldn't find an explanation other than works without issues

File: doom-modeline-core.el [hunk 2]
```diff
@@ -789,7 +789,9 @@ then this function does nothing."
   "Cache the font width."
   (let ((attributes (face-all-attributes 'mode-line)))
     (or (cdr (assoc attributes doom-modeline--font-width-cache))
-        (let ((width (window-font-width nil 'mode-line)))
+        (let ((width (if (and (fboundp 'server-running-p) (server-running-p)) 
+                              (window-width)
```
Comment (seagle0128) [inline doom-modeline-core.el:793]: Why use window width? should be font width here.

Comment (Dr-Dd) [reply doom-modeline-core.el:793]: As i've said before, due to implementation details of the emacs client server protocol and the `windows.el` file, when creating an emacs frame with `emacsclient -c`, the function `window-font-width` returns the wrong window width, while `window-width` in this case works, and returns always the right value. 

On the other hand, when emacs is run with the command `emacs`, with no server running, `window-width` does not return the correct window width, instead, `window-font-width` does.  

So, in other words, the logic behind this is (in pseudocode), is:

````C
if (server_is_running) {
  width = window-width;
} else { // The server is not running, emacs is a standalone instance
  width = window-font-width;
}
````

Comment (seagle0128) [reply doom-modeline-core.el:793]: it doesn't make sense to me. The font width is what we need here, while not window width. In daemon mode, the font width returns 1 always, that's why re-calculate the font width in `after-make-frame-hook`.

Comment (Dr-Dd) [reply doom-modeline-core.el:793]: I've done a little bit of digging. **Changing any fonts** as described in the [README.md](https://github.com/seagle0128/doom-modeline/blob/master/README.md#faq) (either via `set-face-attributes` or `custom-set-faces`) **breakes the bar alignment** when running `emacsclient -c` (see #354). The reason behind this is `doom-modeline-refresh-font-width-cache` isn't called late enough in the init process if emacs is running as a daemon. In fact, calling `doom-modeline-refresh-font-width-cache` manually after opening a frame fixes the modeline alignment for that frame. 

Now, 2 solutions come to mind:

- Rework the way `doom-modeline--font-width-cache` is updated so that it returns the correct width even when opening an `emacsclient -c` instance (the current solution involving `window-setup-hook`, `after-make-frame-functions` and `after-setting-font-hook` **doesn't** fix the problem)
- Using this pull-request, modifying it so that it is conforms to your code standards

Solving this bug would enable medium-size fonts users like me to use this modeline without making it occupy a large part of the screen, letting people change font size and family as they see fit.

Comment (seagle0128) [reply doom-modeline-core.el:793]: Regarding to solution 2, it's not related to code standards. It should return the width of a character, while not the window width. The fix is invalid IMO.
